### ISSUE TITLE
Initial Migrations do no longer work

### ISSUE THREAD

**Issue by l0nax** [uptrace/uptrace#73]
#### Description

If a new deployment is created with the latest `uptrace-dev` docker image, the initial Clickhouse migrations will not pass.

This is because the Clickhouse formatter does not fully replace all arguments within the migrations.

The following lines declare which arguments do exist
https://github.com/uptrace/uptrace/blob/93bb17c7b46c4d1dc721073b948bd1fb80b3a4af/pkg/bunapp/app.go#L363-L370

But the resulting query looks like this:
```sql
CREATE TABLE uptrace.spans_index ?ON_CLUSTER (
  project_id UInt32 Codec(DoubleDelta, ?CODEC),
  system LowCardinality(String) Codec(?CODEC),
  group_id UInt64 Codec(Delta, ?CODEC),

  trace_id UUID Codec(?CODEC),
  id UInt64 Codec(?CODEC),
  parent_id UInt64 Codec(?CODEC),
  name LowCardinality(String) Codec(?CODEC),
  event_name String Codec(?CODEC),
  is_event UInt8 ALIAS event_name != '',
  kind LowCardinality(String) Codec(?CODEC),
  time DateTime Codec(Delta, ?CODEC),
  duration Int64 Codec(Delta, ?CODEC),
  count Float32 Codec(?CODEC),

  status_code LowCardinality(String) Codec(?CODEC),
  status_message String Codec(?CODEC),

  link_count UInt8 Codec(?CODEC),
  event_count UInt8 Codec(?CODEC),
  event_error_count UInt8 Codec(?CODEC),
  event_log_count UInt8 Codec(?CODEC),

  all_keys Array(LowCardinality(String)) Codec(?CODEC),
  attr_keys Array(LowCardinality(String)) Codec(?CODEC),
  attr_values Array(String) Codec(?CODEC),

  "service.name" LowCardinality(String) Codec(?CODEC),
  "host.name" LowCardinality(String) Codec(?CODEC),

  "db.system" LowCardinality(String) Codec(?CODEC),
  "db.statement" String Codec(?CODEC),
  "db.operation" LowCardinality(String) Codec(?CODEC),
  "db.sql.table" LowCardinality(String) Codec(?CODEC),

  "log.severity" LowCardinality(String) Codec(?CODEC),
  "log.message" String Codec(?CODEC),

  "exception.type" LowCardinality(String) Codec(?CODEC),
  "exception.message" String Codec(?CODEC),

  INDEX idx_attr_keys attr_keys TYPE bloom_filter(0.01) GRANULARITY 64,
  INDEX idx_duration duration TYPE minmax GRANULARITY 1
)
ENGINE = ?REPLICATEDMergeTree()
ORDER BY (project_id, system, group_id, time)
PARTITION BY toDate(time)
TTL toDate(time) + INTERVAL ?SPANS_TTL DELETE
SETTINGS ttl_only_drop_parts = 1,
         storage_policy = ?SPANS_STORAGE
```

Only the first argument `?DB` is replaced correctly.

**Comment by vmihailenco** [comment]
@l0nax thanks for reporting, but it seems to be working for me.

What I tried:

```shell
cd example/docker
docker-compose pull
docker-compose up -d
```

Are there any details that I am missing?

>The following lines declare which arguments do exist

That code is slightly outdated, but I am not sure it is relevant. The latest version is at https://github.com/uptrace/uptrace/blob/master/cmd/uptrace/command/ch.go#L264-L296

**Comment by l0nax** [comment]
I tried to use the `docker-compose.yaml` and the following error appierd in the logs:
```
docker-uptrace-1         | migrated to group #1 (20220820073857)
docker-uptrace-1         | [ch]  11:53:49.051   CREATE                1.197ms  CREATE TABLE uptrace.spans_index ?ON_CLUSTER (
docker-uptrace-1         |   project_id UInt32 Codec(DoubleDelta, ?CODEC),
docker-uptrace-1         |   system LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   group_id UInt64 Codec(Delta, ?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   trace_id UUID Codec(?CODEC),
docker-uptrace-1         |   id UInt64 Codec(?CODEC),
docker-uptrace-1         |   parent_id UInt64 Codec(?CODEC),
docker-uptrace-1         |   name LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   event_name String Codec(?CODEC),
docker-uptrace-1         |   is_event UInt8 ALIAS event_name != '',
docker-uptrace-1         |   kind LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   time DateTime Codec(Delta, ?CODEC),
docker-uptrace-1         |   duration Int64 Codec(Delta, ?CODEC),
docker-uptrace-1         |   count Float32 Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   status_code LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   status_message String Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   link_count UInt8 Codec(?CODEC),
docker-uptrace-1         |   event_count UInt8 Codec(?CODEC),
docker-uptrace-1         |   event_error_count UInt8 Codec(?CODEC),
docker-uptrace-1         |   event_log_count UInt8 Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   all_keys Array(LowCardinality(String)) Codec(?CODEC),
docker-uptrace-1         |   attr_keys Array(LowCardinality(String)) Codec(?CODEC),
docker-uptrace-1         |   attr_values Array(String) Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   "service.name" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   "host.name" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   "db.system" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   "db.statement" String Codec(?CODEC),
docker-uptrace-1         |   "db.operation" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   "db.sql.table" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   "log.severity" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   "log.message" String Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   "exception.type" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   "exception.message" String Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   INDEX idx_attr_keys attr_keys TYPE bloom_filter(0.01) GRANULARITY 64,
docker-uptrace-1         |   INDEX idx_duration duration TYPE minmax GRANULARITY 1
docker-uptrace-1         | )
docker-uptrace-1         | ENGINE = ?REPLICATEDMergeTree()
docker-uptrace-1         | ORDER BY (project_id, system, group_id, time)
docker-uptrace-1         | PARTITION BY toDate(time)
docker-uptrace-1         | TTL toDate(time) + INTERVAL ?SPANS_TTL DELETE
docker-uptrace-1         | SETTINGS ttl_only_drop_parts = 1,
docker-uptrace-1         |          storage_policy = ?SPANS_STORAGE
docker-uptrace-1         |
docker-uptrace-1         |        *ch.Error: DB::Exception: Syntax error: failed at position 34 ('?') (line 1, col 34): ?ON_CLUSTER (
docker-uptrace-1         |   project_id UInt32 Codec(DoubleDelta, ?CODEC),
docker-uptrace-1         |   system LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   group_id UInt64 Codec(Delta, ?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   trace_id. Expected one of: token, Dot, UUID, ON, OpeningRoundBracket, storage definition, ENGINE, PARTITION BY, PRIMARY KEY, ORDER BY, SAMPLE BY, TTL, EMPTY AS, AS, COMMENT, INTO OUTFILE, FORMAT, SETTINGS, end of query
docker-uptrace-1         | 2022-08-29T11:53:49.168Z     error   uptrace/main.go:143     ClickHouse migrations failed    {"error": "DB::Exception: Syntax error: failed at position 34 ('?') (line 1, col 34): ?ON_CLUSTER (\n  project_id UInt32 Codec(DoubleDelta, ?CODEC),\n  system LowCardinality(String) Codec(?CODEC),\n  group_id UInt64 Codec(Delta, ?CODEC),\n\n  trace_id. Expected one of: token, Dot, UUID, ON, OpeningRoundBracket, storage definition, ENGINE, PARTITION BY, PRIMARY KEY, ORDER BY, SAMPLE BY, TTL, EMPTY AS, AS, COMMENT, INTO OUTFILE, FORMAT, SETTINGS, end of query"}
docker-uptrace-1         | main.glob..func2
docker-uptrace-1         |      github.com/uptrace/uptrace/cmd/uptrace/main.go:143
docker-uptrace-1         | github.com/urfave/cli/v2.(*Command).Run
docker-uptrace-1         |      github.com/urfave/cli/v2@v2.11.1/command.go:169
docker-uptrace-1         | github.com/urfave/cli/v2.(*App).RunContext
docker-uptrace-1         |      github.com/urfave/cli/v2@v2.11.1/app.go:378
docker-uptrace-1         | github.com/urfave/cli/v2.(*App).Run
docker-uptrace-1         |      github.com/urfave/cli/v2@v2.11.1/app.go:251
docker-uptrace-1         | main.main
docker-uptrace-1         |      github.com/uptrace/uptrace/cmd/uptrace/main.go:71
docker-uptrace-1         | runtime.main
docker-uptrace-1         |      runtime/proc.go:250
docker-uptrace-1         | 2022-08-29T11:53:49.211Z     info    tracing/span_processor.go:47    starting processing spans...    {"threads": 24, "batch_size": 24000, "buffer_size": 1152000}
docker-uptrace-1         | 2022-08-29T11:53:49.215Z     info    metrics/measure_processor.go:64 starting processing metrics...  {"threads": 24, "batch_size": 24000, "buffer_size": 1152000}
```

Record of the terminal session with the error: https://asciinema.org/a/A3lbwCIrc2wuYuKFHYab7e0xp

**Comment by vmihailenco** [comment]
@l0nax Could you try the latest docker example and v1.0.0-rc.2?

**Comment by vmihailenco** [comment]
@l0nax could show `docker-compose logs uptrace`?

**Comment by l0nax** [comment]
Sure:
```
docker-uptrace-1         | 2022-08-29T13:00:31.486Z     info    uptrace/main.go:114     starting Uptrace...     {"version": "v1.0.0-rc.2", "config": "/etc/uptrace/uptrace.yml"}
docker-uptrace-1         | migrated to group #1 (20220820073857)
docker-uptrace-1         | [ch]  13:00:31.763   CREATE                 1.45ms  CREATE TABLE uptrace.spans_index ?ON_CLUSTER (
docker-uptrace-1         |   project_id UInt32 Codec(DoubleDelta, ?CODEC),
docker-uptrace-1         |   system LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   group_id UInt64 Codec(Delta, ?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   trace_id UUID Codec(?CODEC),
docker-uptrace-1         |   id UInt64 Codec(?CODEC),
docker-uptrace-1         |   parent_id UInt64 Codec(?CODEC),
docker-uptrace-1         |   name LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   event_name String Codec(?CODEC),
docker-uptrace-1         |   is_event UInt8 ALIAS event_name != '',
docker-uptrace-1         |   kind LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   time DateTime Codec(Delta, ?CODEC),
docker-uptrace-1         |   duration Int64 Codec(Delta, ?CODEC),
docker-uptrace-1         |   count Float32 Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   status_code LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   status_message String Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   link_count UInt8 Codec(?CODEC),
docker-uptrace-1         |   event_count UInt8 Codec(?CODEC),
docker-uptrace-1         |   event_error_count UInt8 Codec(?CODEC),
docker-uptrace-1         |   event_log_count UInt8 Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   all_keys Array(LowCardinality(String)) Codec(?CODEC),
docker-uptrace-1         |   attr_keys Array(LowCardinality(String)) Codec(?CODEC),
docker-uptrace-1         |   attr_values Array(String) Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   "service.name" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   "host.name" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   "db.system" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   "db.statement" String Codec(?CODEC),
docker-uptrace-1         |   "db.operation" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   "db.sql.table" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   "log.severity" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   "log.message" String Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   "exception.type" LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   "exception.message" String Codec(?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   INDEX idx_attr_keys attr_keys TYPE bloom_filter(0.01) GRANULARITY 64,
docker-uptrace-1         |   INDEX idx_duration duration TYPE minmax GRANULARITY 1
docker-uptrace-1         | )
docker-uptrace-1         | ENGINE = ?REPLICATEDMergeTree()
docker-uptrace-1         | ORDER BY (project_id, system, group_id, time)
docker-uptrace-1         | PARTITION BY toDate(time)
docker-uptrace-1         | TTL toDate(time) + INTERVAL ?SPANS_TTL DELETE
docker-uptrace-1         | SETTINGS ttl_only_drop_parts = 1,
docker-uptrace-1         |          storage_policy = ?SPANS_STORAGE
docker-uptrace-1         |
docker-uptrace-1         |        *ch.Error: DB::Exception: Syntax error: failed at position 34 ('?') (line 1, col 34): ?ON_CLUSTER (
docker-uptrace-1         |   project_id UInt32 Codec(DoubleDelta, ?CODEC),
docker-uptrace-1         |   system LowCardinality(String) Codec(?CODEC),
docker-uptrace-1         |   group_id UInt64 Codec(Delta, ?CODEC),
docker-uptrace-1         |
docker-uptrace-1         |   trace_id. Expected one of: token, Dot, UUID, ON, OpeningRoundBracket, storage definition, ENGINE, PARTITION BY, PRIMARY KEY, ORDER BY, SAMPLE BY, TTL, EMPTY AS, AS, COMMENT, INTO OUTFILE, FORMAT, SETTINGS, end of query
docker-uptrace-1         | 2022-08-29T13:00:31.781Z     error   uptrace/main.go:146     ClickHouse migrations failed    {"error": "DB::Exception: Syntax error: failed at position 34 ('?') (line 1, col 34): ?ON_CLUSTER (\n  project_id UInt32 Codec(DoubleDelta, ?CODEC),\n  system LowCardinality(String) Codec(?CODEC),\n  group_id UInt64 Codec(Delta, ?CODEC),\n\n  trace_id. Expected one of: token, Dot, UUID, ON, OpeningRoundBracket, storage definition, ENGINE, PARTITION BY, PRIMARY KEY, ORDER BY, SAMPLE BY, TTL, EMPTY AS, AS, COMMENT, INTO OUTFILE, FORMAT, SETTINGS, end of query"}
docker-uptrace-1         | main.glob..func2
docker-uptrace-1         |      github.com/uptrace/uptrace/cmd/uptrace/main.go:146
docker-uptrace-1         | github.com/urfave/cli/v2.(*Command).Run
docker-uptrace-1         |      github.com/urfave/cli/v2@v2.11.1/command.go:169
docker-uptrace-1         | github.com/urfave/cli/v2.(*App).RunContext
docker-uptrace-1         |      github.com/urfave/cli/v2@v2.11.1/app.go:378
docker-uptrace-1         | github.com/urfave/cli/v2.(*App).Run
docker-uptrace-1         |      github.com/urfave/cli/v2@v2.11.1/app.go:251
docker-uptrace-1         | main.main
docker-uptrace-1         |      github.com/uptrace/uptrace/cmd/uptrace/main.go:71
docker-uptrace-1         | runtime.main
docker-uptrace-1         |      runtime/proc.go:250
docker-uptrace-1         | 2022-08-29T13:00:31.823Z     info    tracing/span_processor.go:47    starting processing spans...    {"threads": 24, "batch_size": 24000, "buffer_size": 1152000}
docker-uptrace-1         | 2022-08-29T13:00:31.826Z     info    metrics/measure_processor.go:64 starting processing metrics...  {"threads": 24, "batch_size": 24000, "buffer_size": 1152000}
docker-uptrace-1         | 2022-08-29T13:00:31.827Z     info    uptrace/main.go:361     starting monitoring metrics...  {"rules": 4}
```

**Comment by vmihailenco** [comment]
This should be fixed in v1.0.0-rc3. Thanks for the report!
